sdk/storage/azure-storage-quickquery/src/main/java/com/azure/storage/quickquery/implementation/util/FluxInputStream.java

```diff
@@ -0,0 +1,220 @@
+// Licensed under the MIT License.
+
+package com.azure.storage.quickquery.implementation.util;
+
+import com.azure.core.util.FluxUtil;
+import com.azure.core.util.logging.ClientLogger;
+import com.azure.storage.blob.models.BlobStorageException;
+import com.azure.storage.common.implementation.Constants;
+import org.reactivestreams.Subscription;
+import reactor.core.publisher.Flux;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+/**
+ * An InputStream interface that subscribes to a Flux allows data to be read.
+ */
+public class FluxInputStream extends InputStream {
+
+    private ClientLogger logger;
+
+    private Flux<ByteBuffer> data;
+
+    private Subscription subscription; // Subscription to request more data from as needed
+
+    private ByteArrayInputStream buffer;
+
+    private boolean subscribed;
+    private boolean fluxComplete;
+    private boolean waitingForData;
+
+    private final Lock lock;
+    private final Condition dataAvailable;
+
+    private IOException lastError;
+
+    /**
+     * Creates a new FluxInputStream
+     *
+     * @param data The data to subscribe to and read from.
+     * @param logger {@link ClientLogger}
+     */
+    public FluxInputStream(Flux<ByteBuffer> data, ClientLogger logger) {
+        this.subscribed = false;
+        this.fluxComplete = false;
+        this.waitingForData = false;
+        this.data = data;
+        this.logger = logger;
+        this.lock = new ReentrantLock();
+        this.dataAvailable = lock.newCondition();
+    }
+
+    @Override
+    public int read() throws IOException {
+        return read(new byte[1], 0, 1);
+    }
+
+    @Override
+    public int read(byte[] b, int off, int len) throws IOException {
+        validateParameters(b, off, len);
+
+        /* If len is 0, then no bytes are read and 0 is returned. */
+        if (len == 0) {
+            return 0;
+        }
+        /* Attempt to read at least one byte. If no byte is available because the stream is at end of file,
+           the value -1 is returned; otherwise, at least one byte is read and stored into b. */
+
+        /* Not subscribed? subscribe and block for data */
+        if (!subscribed) {
+            blockForData();
+        }
+        /* Now, we have subscribed. */
+        /* At this point, buffer should not be null. If it is, that indicates either an error or completion event
+           was emitted by the Flux. */
+        if (this.buffer == null) { // Only executed on first subscription.
+            if (this.lastError != null) {
+                throw lastError;
+            }
+            if (this.fluxComplete) {
+                return -1;
+            }
+            this.buffer = new ByteArrayInputStream(new byte[0]);
```

Comment:
What's the purpose of this line?

Reply:
That whole if block is really a "this is no longer the first read check." InputStreams are supposed to throw if an error happens on the first read.

Reply:
So I make buffer non null once the first read is done